xdsl/backend/asm_perf_reporter.py

```diff
@@ -0,0 +1,81 @@
+import os
+import subprocess
+from abc import ABC, abstractmethod
+from shutil import which
+from tempfile import NamedTemporaryFile
+
+from xdsl.backend.assembly_printer import AssemblyPrintable, AssemblyPrinter
+from xdsl.ir import Block
+
+
+class AssemblyPerformanceReporter(ABC):
+    arch: str
+    src_path: str
+
+    def __init__(self, arch: str):
+        self.arch = arch
+
+    @abstractmethod
+    def name(self) -> str:
+        pass
+
+    @abstractmethod
+    def cmd(self) -> list[str]:
+        pass
+
+    def produce_report(self) -> str:
+        cmd = self.cmd()
+        with NamedTemporaryFile(mode="w+", suffix=".s", delete=False) as report_file:
+            _ = subprocess.run(
+                cmd, stdout=report_file, stderr=subprocess.DEVNULL, check=True
+            )
+            report_file.seek(0)
+            return report_file.read()
+
+    @abstractmethod
+    def process_report(self, report: str) -> float | None:
+        pass
+
+    def is_installed(self) -> bool:
+        return which(self.name()) is not None
+
+    def estimate_throughput(self, block: Block) -> float | None:
+        with NamedTemporaryFile(mode="w+", delete=False, suffix=".s") as tmp_file:
+            self.src_path = tmp_file.name
+            printer = AssemblyPrinter(stream=tmp_file)
+            for op in block.walk():
+                assert isinstance(op, AssemblyPrintable), (
+                    f"Block operation {op} should be an assembly instruction"
+                )
+                op.print_assembly(printer)
+
+        try:
+            report = self.produce_report()
+            return self.process_report(report)
+        finally:
+            if os.path.exists(self.src_path):
+                os.remove(self.src_path)
```

Comment:
One day, I hope to implement a block throughput estimator in xDSL. When that day comes, most of this code won't be compatible, since it assumes that the estimation will be done via a command-line call. Can you please move it into the MCAReporter, and leave only `estimate_throughput` as an abstract function?

Can you please also add a doc string to the file, the class, and the function?